### PR TITLE
Decoupled Weight Decay Regularization

### PR DESCRIPTION
**Pull request by jingweiz**
Hey,
We added `SGDW` and `AdamW` in `optim`, accoridng to the new ICLR submission from Loshchilov and Hutter: [Decoupled Weight Decay Regularization
](https://arxiv.org/abs/1711.05101).
We also found some inconsistency of the current implementation of `Adam` with the original Adam paper and we asked about it here: [https://discuss.pytorch.org/t/adam-implementation/10031](https://discuss.pytorch.org/t/adam-implementation/10031). We tried to make it consistent in the commit: `f195087a3666f85a417ee7561bec439bb68f81c3`, but then it failed during testing which might due to some inconsistency with the `legacy.optim.adam` so we changed it back in the last commit. But it would great if you guys can check it to see if this is actually an inconsistency then we can add it back.
Thanks in advance!

### CODE REVIEW

**Comment by radekosmulski** [comment]
There seems to exist `sgdw.py` in the root of the directory. Not sure if this was intentional?

The docstring for SGDRCosineLR in the example section seems to be referencing `CosineLR` and also the parameter names seem off. 

In the step method, we call get_lr but don't pass it epoch_idx which seems to be a required param (I tried this and it blows up).

Thank you for working on this :+1: Happy to help in whatever capacity I can - will continue to test this as new versions become available.

**Comment by TiRune** [comment]
I tested the current code on a simple SeNET network on CIFAR10. The current ADAMW code I found here works nearly identically to SGD with momentum 0.9, and does not seem to decay the learning rate at all. In contract to that, the original ADAM optimiser decays the learning rate properly and converges very differently from SGD. 

I sincerely wonder if the current ADAMW implementation here is correct.

File: torch/optim/adamw.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                denom = exp_avg_sq.sqrt().add_(group['eps'])
+
+                bias_correction1 = 1 - beta1 ** state['step']
+                bias_correction2 = 1 - beta2 ** state['step']
+                step_size = group['lr'] * math.sqrt(bias_correction2) / bias_correction1
+
+                p.data.addcdiv_(-step_size, exp_avg, denom)
+
+                if group['weight_decay'] != 0:
+                    p.data.add_(-group['weight_decay'], p.data)
```
Comment (colllin) [inline torch/optim/adamw.py]: There might be a mistake here: See that `p.data` in L82 was already updated based on the gradient in L79.  Thus it appears that the weight decay is being multiplied by the weights *after* the gradient-based update, and instead it should be multiplied by the original weights (before the gradient-based update).  I think this can be fixed by simply moving L81-82 above L79, since the `step_size`, `exp_avg`, and `denom` are already computed.  This will compute the weight decay separate from the gradient update.

I found this based on the AdamW description in http://www.fast.ai/2018/07/02/adam-weight-decay/

They also multiply the weight decay by the learning rate (`-wd * lr * w`), and I applied both of these updates to your AdamW implementation here:

https://gist.github.com/colllin/0b146b154c4351f9a40f741a28bff1e3

I'm not sure whether you intentionally computed `-wd * w` instead of `-wd * lr * w` — if so, could you explain why you decided that?

**Comment by episodeyang** [comment]
Just want to bump this PR up.

**Comment by OutSorcerer** [comment]
The fact that the paper was [accepted](https://openreview.net/forum?id=Bkg6RiCqY7) to ICLR 2019 should be also taken in consideration.

**Comment by taohu88** [comment]
BTW, to do it in right way, the weight decay should be multiplied by lr. I saw many proposals didn't do that. This will break how people set weight decay in old way. Suddenly, they have to set weight decay much lower.

**Review by taohu88 — COMMENTED** [review]
to be clear, we should do this
```
                # p = p - wd * lr * p
                # according to the paper, weight decay should be done here
                # be consistent to general approach, the weight decay will be always multiplied by lr
                wd = group['weight_decay']
                if wd != 0:
                    p.data.add_(-wd * lr, p.data)
                
                # note (lr * p.grad) = step_size * exp_avg/denom
                # so p = p - lr * p.grad = p - step_size * exp_avg/denom
                p.data.addcdiv_(-step_size, exp_avg, denom)
                # net result p = p_old - wd * lr * p_old - step_size * exp_avg/denom
```

**Comment by BramVanroy** [comment]
This is labelled as awaiting response, but I'm not sure from whom. From @jingweiz ? Or from torch developers?

**Comment by bhack** [comment]
Any news?

**Comment by naifrec** [comment]
hello, does anyone know if Fixing Weight Decay Regularization in Adam could be the reason why [Keras (tf backend) seem to outperform PyTorch](https://www.reddit.com/r/MachineLearning/comments/7nw67c/d_pytorch_are_adam_and_rmsprop_okay/) (given exact same training pipeline) when using Adam? If so, can we prioritize the merging of this PR?

**Comment by eukaryote31** [comment]
What is this PR waiting on right now?

**Comment by soumith** [comment]
The title might seem like this is fixing a software bug, but it isn't.
The authors propose a change to the Adam algorithm, that performs better under certain conditions.

this is relevant to comments made by @naifrec 

About the awaiting-response tag, it was tagged to get a response from the author on @colllin 's comments on the PR.

**Comment by KushajveerSingh** [comment]
Isn't it a software bug. When we specify L2 regularization pytorch is actually using weight decay instead.

**Comment by DeepAlchemist** [comment]
@taohu88 As the [original paper](https://arxiv.org/pdf/1711.05101.pdf) show, i.e., the green region in Algorithm 2.
图片

it should be   
`if wd != 0:`
      ` p.data.add_(-wd * ScheduleMultiplier, p.data)  ---  (1)`
rather than
`if wd != 0:`
      `p.data.add_(-wd * lr, p.data) --- (2)`
In fact, [Alexnet](https://papers.nips.cc/paper/4824-imagenet-classification-with-deep-convolutional-neural-networks.pdf) achieves weight decay by equation (2). I don't know why p.data in AdamW is not multiplied by the learning rate.

BTW, it seems not straightforward to get the sheduler multiplier of lr_sheduler in pytorch.

**Comment by soumith** [comment]
cc: @vincentqb can you review this and get it to completion if it makes sense. Use the guidelines that I shared with you separately.

**Comment by vincentqb** [comment]
> The title might seem like this is fixing a software bug, but it isn't.
> The authors propose a change to the Adam algorithm, that performs better under certain conditions.

Changed name from "Fixing Weight Decay Regularization in Adam" to "Decoupled Weight Decay Regularization" to reflect the name of the paper being quoted.

File: torch/optim/adamw.py (hunk 1 above)
Comment (henrique) [inline torch/optim/adamw.py]: The class comments says weight_decay == L2 regularisation. Shouldn't this be the square of the parameters then?

```suggestion
                    p.data.addcmul_(-group['weight_decay'], p.data, p.data)
```

either way, thanks a lot for the contribution!

Comment (spott) [reply torch/optim/adamw.py]: Isn't this because the you are subtracting the derivative of the L2 norm of the weights, not the L2 norm itself?

**Comment by farhadrgh** [comment]
Is there any particular reason you are not adding AdamaxW?
